django/db/models/expressions.py

```diff
@@ -651,6 +654,39 @@ def relabeled_clone(self, relabels):
         return self
 
 
+class SliceableF(F):
+    """
+    An object that contains a slice of an F expression.
+
+    Object resolves the column on which the slicing is applied, and then
+    applies the slicing if possible.
+    """
+    def __init__(self, f_obj, slice_obj):
+        if isinstance(slice_obj, int):
+            if slice_obj < 0:
+                raise ValueError('Negative indexing is not supported.')
+            self.start = slice_obj + 1
+            self.length = 1
+        elif isinstance(slice_obj, slice):
+            if (
+                (slice_obj.start is not None and slice_obj.start < 0) or
+                (slice_obj.stop is not None and slice_obj.stop < 0)
+            ):
+                raise ValueError('Negative indexing is not supported.')
+            if slice_obj.step is not None:
+                raise ValueError('Step argument is not supported.')
+            self.start = 1 if slice_obj.start is None else int(slice_obj.start) + 1
+            self.length = None if slice_obj.stop is None else int(slice_obj.stop) - self.start + 1
+        else:
+            raise TypeError('Argument to slice must be either int or slice instance.')
+        self.name = f_obj.name
```

Comment:
I'd move this line to the top of `__init__()` so it isn't lost below all the conditional logic.